Sources/NIO/EventLoopFuture.swift

```diff
@@ -957,6 +957,43 @@ extension EventLoopFuture {
     }
 }
 
+extension EventLoopFuture {
+    /// Returns a new `EventLoopFuture` that succeeds when all of the provided `EventLoopFuture`s complete.
+    /// The new `EventLoopFuture` will contain an array of results, maintaining ordering for each of the `EventLoopFuture`s.
+    ///
+    /// The returned `EventLoopFuture` always succeeds, regardless of any failures from the waiting futures.
+    ///
+    /// If it is desired to flatten them into a single `EventLoopFuture` that fails on the first `EventLoopFuture` failure,
+    /// use one of the `reduce` methods instead.
+    /// - Parameter futures: An array of homogenous `EventLoopFuture`s to gather results from.
+    /// - Returns: A new `EventLoopFuture` with all the results of the provided futures.
+    public static func whenAllComplete<InputValue>(_ futures: [EventLoopFuture<InputValue>],
+                                                   eventLoop: EventLoop) -> EventLoopFuture<[Result<InputValue, Error>]> {
+        return eventLoop.makeSucceededFuture(result: [])
+            .flatMap { _ in
+                let promise = eventLoop.makePromise(of: [Result<InputValue, Error>].self)
+
+                var remainingCount = futures.count
+                var results: [Result<InputValue, Error>?] = .init(repeating: nil, count: futures.count)
+
+                for (index, future) in futures.enumerated() {
+                    future
+                        .hopTo(eventLoop: eventLoop)
+                        ._whenComplete {
+                            results[index] = future.value!
+                            remainingCount -= 1
+
+                            guard remainingCount == 0 else { return CallbackList() }
+
+                            return promise._setValue(value: .success(results.compactMap({ $0 })))
+                        }
+                }
+
+                return promise.futureResult
+            }
+    }
```

Comment:
Rather than use the do-nothing future object so you can get access to `flatMap`, I propose you use `eventLoop.execute` instead. Specifically, let's pull out the body of the `flatMap` call into a separate function (call it `_whenAllComplete0`) and then you can do this:

```swift
let promise = eventLoop.makePromise(of: [Result<InputValue, Error>].self)

if eventLoop.inEventLoop {
    _whenAllComplete0(promise, futures, eventLoop: eventLoop)
} else {
    eventLoop.execute {
        _whenAllComplete0(promise, futures, eventLoop: eventLoop)
    }
}

return promise.futureResult
```

That way we don't need the extra do-nothing future, and we keep the ability to avoid unnecessary thread-hops.

Reply:
Oh, that's much better! This will probably come to me more naturally as I spend more time working in the mindset of making sure things are in the event loop, rather than working just off of attached callbacks on ELFS :)

Reply:
Quick question - what is the meaning behind the `0` suffix? I understand having a `_` prefix to label as SPI and remove ambiguities, but what additional meaning does `0` provide?

Reply:
We use it as an internal style marker to indicate that the function is *not* thread safe. Almost all functions with `0` in their name begin with `eventLoop.assertInEventLoop()` in their body. I recommend doing the same, incidentally.